### PR TITLE
CORE-19845 Tactical fix to stop COMPLETED flows being marked as FAILED

### PR DESCRIPTION
**Pull request by LWogan**
For: 5.2.1

**Problem**:
In order to maintain performance of the corda system, deletion of states in the mediator has been moved to after committing offsets in the bus. 

If the system crashes after offsets are committed but before states are deleted or if the delete call fails then the states can be orphaned and will not be deleted immediately. 

This will then result in Checkpoints being picked up by the idle flow scheduled job which will mark the flow as failed even though it has already updated the status to COMPLETED. This job is currently set to trigger every 30 mins This means a flow would be marked as FAILED 30 mins after being marked as COMPLETED.

**Solution**:
One possible solution would be to update the flow status processor which sets the status in the state manager to ignore updates to status when the status is already at  terminal status (FAILED/KILLED/COMPLETED) 

This will not stop the checkpoint state from being picked up by the idle flow processor job. However for COMPLETED flows all sessions are terminated. The only negative impact of running a COMPLETED state through this job inadvertently is the update to Flow Status. No messages will be sent to counterparties.

### CODE REVIEW

**Comment by LWogan** [comment]
This has been flagged as something we'd like to do for 5.2.1. Raising this PR to highlight a solution.

File: components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt [hunk 1]
```diff
@@ -52,6 +56,16 @@ class DurableFlowStatusProcessor(
         return emptyList()
     }
 
+    private fun getStatus(state: State) : String? {
+        val flowStatus = state.metadata[FLOW_STATUS_METADATA_KEY]
+        return if (flowStatus != null) flowStatus as String else null
+    }
```
Comment (ben-millar) [inline components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt]:
```suggestion
    private fun getStatus(state: State): String? =
        state.metadata[FLOW_STATUS_METADATA_KEY] as? String
```

File: components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt [hunk 2]
```diff
@@ -52,6 +56,16 @@ class DurableFlowStatusProcessor(
...
+
+    private fun hasTerminatedStatus(state: State): Boolean {
+        val status = getStatus(state)
+        return status != null && TERMINATED_STATES.contains(FlowStates.valueOf(status))
+    }
```
Comment (ben-millar) [inline components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt]:
```suggestion
    private fun hasTerminatedStatus(state: State): Boolean =
        getStatus(state)?.let { FlowStates.valueOf(it) in TERMINATED_STATES } ?: false
```

**Review by ben-millar — DISMISSED** [review]
This change makes sense, plus it's adding very little work on top of what we're already doing. LGTM!

File: components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt [hunk 3]
```diff
@@ -28,13 +28,17 @@ class DurableFlowStatusProcessor(
...
+    companion object {
+        private val TERMINATED_STATES = setOf(FlowStates.COMPLETED, FlowStates.KILLED, FlowStates.FAILED)
+    }
+
     override val keyClass: Class<FlowKey> get() = FlowKey::class.java
     override val valueClass: Class<FlowStatus> get() = FlowStatus::class.java
     override fun onNext(events: List<Record<FlowKey, FlowStatus>>): List<Record<*, *>> {
         val flowKeys = events.map { it.key.hash() }
-        val existingStates = stateManager.get(flowKeys)
+        val existingStates = stateManager.get(flowKeys).filterNot { hasTerminatedStatus(it.value) }
```
Comment (mbrkic-r3) [inline components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt]: Assuming there is a new record coming with status "FAILED" and there was already existing status "COMPLETED" in DB, wouldn't this change treat new record as "new state" and try to persist it using the same key (which would most likely fail)?

**Comment by mbrkic-r3** [comment]
Changes in PR are related to status, but what will happen with orphaned checkpoint, would it stay in DB forever?

File: components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt (hunk 3 above)
Comment (LWogan) [reply components/flow/flow-rest-resource-service-impl/src/main/kotlin/net/corda/flow/rest/impl/DurableFlowStatusProcessor.kt]: You are correct. Ive updated the position of the condition

**Comment by LWogan** [comment]
> Changes in PR are related to status, but what will happen with orphaned checkpoint, would it stay in DB forever?

It will get deleted by the idle checkpoint cleanup job. The only output from that job which is undesirable is the update of status to ERROR.

**Review by mbrkic-r3 — APPROVED** [review]
LGTM

**Review by thiagoviana — APPROVED** [review]

**Pull request merged**